src/toil/lib/history_submission.py

```diff
@@ -0,0 +1,677 @@
+
+    :param options: Dict from machine-readable option key to button text.
+    :returns: the key of the selected option, or None if the user declined to
+        select an option.
+    :raises: an exception if the dialog cannot be displayed.
+    """
+
+    # See
+    # <https://python-prompt-toolkit.readthedocs.io/en/master/pages/dialogs.html#button-dialog>
+    # for the dialog and
+    # <https://python-prompt-toolkit.readthedocs.io/en/master/pages/advanced_topics/input_hooks.html>
+    # for how we run concurrently with it.
+
+    from prompt_toolkit.shortcuts import button_dialog
+    from prompt_toolkit.eventloop.inputhook import set_eventloop_with_inputhook
+
+    # We take button options in the reverse order from how prompt_toolkit does it.
+    # TODO: This is not scrollable! What if there's more than 1 screen of text?
+    # TODO: Use come kind of ScrollablePane like <https://stackoverflow.com/q/68369073>
+    application = button_dialog(
+        title=title,
+        text=text,
+        buttons=[(v, k) for k, v in options.items()],
+    )
+
+    # Make the coroutine to run the dialog
+    application_coroutine = application.run_async()
+
+    async def exit_application_after_timeout() -> None:
+        """
+        Wait for the timeout to elapse and then close the application.
+        """
+        await asyncio.sleep(timeout)
+        # End the application and make it return None
+        application.exit()
+
+    # Make the coroutine to stop the dialog
+    timeout_coroutine = exit_application_after_timeout()
+
+    # To avoid messing around with a global event loop and instead just race
+    # some coroutines, we make our own event loop.
+    loop = asyncio.new_event_loop()
+
+    # Attach the coroutines to it so it tries to advance them
+    application_task = loop.create_task(application_coroutine)
+    timeout_task = loop.create_task(timeout_coroutine)
+
+    # This task finishes when the first task in the set finishes.
+    race_task = asyncio.wait({application_task, timeout_task}, return_when=asyncio.FIRST_COMPLETED)
+    # Run it until then
+    loop.run_until_complete(race_task)
+    # Maybe the application needs to finish its exit?
+    loop.run_until_complete(application_task)
+
+    # We no longer need the stopping task/coroutine
+    timeout_task.cancel()
+    # We don't need to await it because we cancel it.
+    # It is OK to cancel it if it is finished already.
+
+    return application_task.result()
+
+# Define the dialog form in the abstract
+Decision = Union[Literal["all"], Literal["current"], Literal["no"], Literal["never"]]
+
+DIALOG_TITLE = "Publish Workflow Metrics on Dockstore?"
+
+# We need to fromat the default config path in to the message, but we can't get
+# it until we can close the circular import loop. So it's a placeholder here.
+# We also leave this un-wrapped to let the dialog wrap it if needed.
+DIALOG_TEXT = """
+Would you like to publish execution metrics to Dockstore?
```

Comment:
Suggestion: replace Dockstore with `dockstore.org` in case it saves someone a step Googling "Dockstore".